Complete content for RStudio IDE docs (#12075)
rstudio/rstudio: docs/user/rstudio/ide/guide/tools/jobs.qmd

title: "RStudio Jobs"
---

RStudio has the ability to send long running R scripts to [local and remote background jobs](https://www.rstudio.com/blog/rstudio-1-2-jobs/). This functionality can dramatically improve the productivity of data scientists and analysts using R since they can continue working in RStudio while jobs are running in the background.

Local background jobs are supported by all versions of the RStudio IDE, server and desktop. Remote background jobs are a feature of [Posit Workbench](https://www.rstudio.com/products/workbench/) and are orchestrated by the [Launcher](https://docs.rstudio.com/job-launcher/), which also supports running interactive sessions on remote resource managers like [Kubernetes](https://kubernetes.io/). More information about setting up and administering the Launcher can be found [here](https://solutions.rstudio.com/launcher/overview).

Background jobs can be started manually or programatically. Local background jobs are ideal for interactive, ad-hoc usage.

## Background jobs

A "background job" is an R script that runs in a separate, dedicated R session. Any R script can be run in a separate session by:

> Selecting "Start Background Job" from the **Background Job** tab in the **Console**

Author: changed

![](images/background-job-console-start.png){fig-alt="Screenshot of the 'Start Background job' tab in the terminal pane."}

> Or from within a .R file by pulling down the Source menu and choosing **Source as Background Job**.

Reviewer: ```suggestion
Or from within a .R file by clicking the **Source** menu and selecting **Source as Background Job**.
```

Author: Changed

![](images/source-job.png){fig-alt="Source script as background job" width="450"}

Either method will give you some options for running your job.

![](images/background-job-option.png){fig-alt="Dialog showing options for starting R script job"}

By default, the job will run in a clean R session, and its temporary workspace will be discarded when the job is complete. This is the fastest and safest configuration, good for reproducible scripts that have no side effects.

However, if you want to feed data from your current R session into the job, or have the job return data to your current R session, change the dialog options as follows:

> -   **Run job with copy of global environment**: If ticked, this option saves your global environment and loads it into the job's R session before it runs. This is useful because it will allow your job to see all the same variables you can see in the IDE. Note that this can be slow if you have large objects in your environment.

Reviewer: ```suggestion
-   **Run job with copy of global environment**: If selected, this option saves your global environment and loads it into the job's R session before it runs. This is useful because it will allow your job to see all the same variables you can see in the IDE. Note that this can be slow if you have large objects in your environment.
```

Author: Changed

-   **Copy job results**: By default, the temporary workspace in which the job runs is not saved. If you'd like to import data from your job back into your R session, you have a couple of choices:

>   -   **Global environment**: This places all the R objects your job creates back in your R session's global environment. Use this option with caution! The objects created by the job will overwrite, without a warning, any objects that have the same name in your environment.

Reviewer: I may nest a callout here vs. having the "Use this option with caution" inline.

```suggestion
  -   **Global environment**: This places all the R objects your job creates back in your R session's global environment. 
  
  :::{.callout-caution}
  Use this option with caution! The objects created by the job will overwrite, without a warning, any objects that have the same name in your environment.
```

Author: added callout

```markdown
# Apply the model ----
```

When your job reaches that line in your script, the name of the section will appear on the progress bar.

![](images/background-job-name.png){fig-alt="Job progress bar showing section progress"}

You can also emit output using the usual R mechanisms, such as:
- `print`
- `message`
- `cat`

This output appears in the Jobs pane when you select your job.

![](images/background-job-print.png){fig-alt="Jobs pane with output of a local job"}

### Scripting

You can script the creation of jobs using the **rstudioapi** package method [jobRunScript](https://www.rdocumentation.org/packages/rstudioapi/versions/0.9.0/topics/jobRunScript); it has options which correspond to each dialog option above. This makes it possible to automate and orchestrate more complicated sets of background tasks.

> Note however that the IDE's background job runner is generally designed for interactive script runs. If you are writing R code and need to run a subtask asynchronously in a background R session, we recommend using the [callr package](https://callr.r-lib.org/) instead.

Reviewer: ```suggestion
However, the IDE's background job runner is generally designed for interactive script runs. If you are writing R code and need to run a subtask asynchronously in a background R session, we recommend using the [callr package](https://callr.r-lib.org/) instead.
```

Author: Changed.

![](images/source-job.png){fig-alt="Source script as background job" width="450"}

Either method will give you some options for running your job.

![](images/background-job-option.png){fig-alt="Dialog showing options for starting R script job"}

By default, the job will run in a clean R session, and its temporary workspace will be discarded when the job is complete. This is the fastest and safest configuration, good for reproducible scripts that have no side effects.

However, if you want to feed data from your current R session into the job, or have the job return data to your current R session, change the dialog options as follows:

-   **Run job with copy of global environment**: If selected, this option saves your global environment and loads it into the job's R session before it runs. This is useful because it will allow your job to see all the same variables you can see in the IDE. Note that this can be slow if you have large objects in your environment.

-   **Copy job results**: By default, the temporary workspace in which the job runs is not saved. If you'd like to import data from your job back into your R session, you have a couple of choices:

> 

Reviewer 2: <img width="795" alt="image" src="https://user-images.githubusercontent.com/114421526/194144903-2c8336c7-733a-4f02-a741-7b95e0682723.png">

should **Global environments** and **Results object** appear as subitems under **Copy jobs results**?

Author: Moved these under level 4 headers so that it is more clear they are a sub section of copy job results.

  
::: {.callout-caution}

Use **Copy jobs results** to **Global environment** option with caution! The objects created by the job will overwrite, without a warning, any objects that have the same name in your environment.

:::

  -   **Results object**: This places all the R objects your job creates into a new environment named `yourscript_results`.

An additional option is to include specific R code to save out the results to files on disk such as `readr::save_rds("results.rds")`

### Lifetime

Background jobs run as non-interactive child R processes of your main R process, which means that they will be shut down if R is. While the R session is running jobs:

-   You will be warned if you attempt to close the window while jobs are still running (on RStudio Desktop)

-   Your R session will not be suspended (on Posit Workbench)

While background jobs are running, a progress bar will appear in the R console summarizing the progress of all running jobs.

![](images/background-job-console.png){fig-alt="R console pane showing job progress tab" width="350"}

### Detailed progress

> The progress bar RStudio shows for your job represents the execution of each top-level statement in your R script. If you want a little more insight into which part of the script is currently running, you can use RStudio's [code sections](https://support.rstudio.com/hc/en-us/articles/200484568-Code-Folding-and-Sections) feature. Add a section marker like this to your R script:

Reviewer 2: Change code sections link to [updated page](https://colorado.rstudio.com/rsc/rstudio-user-guide-2022-09-29/ide/guide/code/code-sections.html)

Author: Changed - thanks!

| [`jobAddOutput()`](https://rstudio.github.io/rstudioapi/reference/jobAddOutput.html)     |   Add Background Job Output  |
| [`jobAddProgress()`](https://rstudio.github.io/rstudioapi/reference/jobAddProgress.html) |  Add Background Job Progress   |
| [`jobRemove()`](https://rstudio.github.io/rstudioapi/reference/jobRemove.html)           |   Remove a Background Job  |
| [`jobSetProgress()`](https://rstudio.github.io/rstudioapi/reference/jobSetProgress.html) |  Set Background Job Progress   |
| [`jobSetState()`](https://rstudio.github.io/rstudioapi/reference/jobSetState.html)       |  Set Background Job State   |
| [`jobSetStatus()`](https://rstudio.github.io/rstudioapi/reference/jobSetStatus.html)     |   Set Background Job Status  |

## Workbench (Launcher) jobs {.pro-header}

On Posit Workbench, you also have the option of running your R script on your company's compute infrastructure, using the [Job Launcher](https://blog.rstudio.com/2018/11/05/rstudio-rsp-1.2-features/). To do this, select:

![](images/workbench-job.png){fig-alt="Source script as launcher job" width="350"}

> See more documentation specific to Workbench Launcher Jobs at <https://docs.posit.co/ide/server-pro/user>

Reviewer 2: This link doesn't open for me

Author: This is a future link - for now I've moved it back to <https://docs.rstudio.com/ide/server-pro/> while we wait for the Workbench User Guide to be finalized.
